man/man1/timeout.1
.TH timeout 1 "19 Jul 2001" 2.0.0 Netatalk 
.SH NAME
timeout \- Send a signal to a program after a certain time
.SH SYNTAX
timeout [\-s \fIsignal\fR] \fIseconds\fR \fIprogram\fR [\fIargs\fR]
.SH DESCRIPTION
\fItimeout\fR executes a \fIprogram\fR (with arguments \fIargs\fR)
and sends a \fIsignal\fR to it after a certain
amount of \fIseconds\fR.
.SH OPTIONS
.TP 
\fB\-s\fR \fIsignal\fR
Signal to send to the spawned process. This can be a numerical
or symbolic ID. This defaults to \fBTERM\fR.
.SH EXAMPLES
.TP 
timeout 10 pap foo.ps
Execute "pap foo.ps" and send a SIGTERM
if \fIpap\fR doesn't return after 10
seconds.
.TP 
timeout \-s HUP 60 sh
Spawn a shell and send a hangup signal after one minute.
.TP 
timeout \-s 9 10 evilprog
Execute a program and \fIKILL\fR it
if it doesn't quit after 10 seconds.



NEWS
Changes in 2.0.0
================

* UPD: afpd: add an error message if -ipaddr parameter cannot be parsed
* UPD: updated documentation
* FIX: afpd: fix a file descriptor and memory leak with OSX ._ resource fork
* FIX: afpd: Prevent overwriting a file by renaming a file in the same 
       directory to the same name. Won't work with OSX, the dest file gets 
       deleted by OSX first.
* FIX: sometimes '0' was used instead of 0 for creator/type
* FIX: removed setpgrp check from configure, we don't use it anymore and
       it doesn't work with cross compile.
* FIX: fix for Solaris "make maintainer-clean", from Alexander Barton
* FIX: fix username matching bug in afppasswd. from kanai at nadmin dot org
* FIX: reworked username check a little. Depending on the UAM, the wrong
       username _could_ have been selected.

Changes in 2.0-rc2
==================

* UPD: use 0 0 for default creator/type rather than UNIX TEXT, from
       Shlomi Yaakobovich.
* UPD: updated documentation
* UPD: change machine type from Macintosh to Netatalk in status reply
* FIX: afpd: CopyFile only create a resource fork for destination if source
       has one.
* FIX: afpd: mangling: for utf8 --> max filename length is 255 bytes, else 31.
* FIX: cnid_dbd: fix a signed/unsigned, 16/32 bits mismatch. from Burkhard
       Schmidt, bs at cpfs.mpg.de.
* FIX: afpd: After ad_setid don't flush resource fork if it has not been
       modified.
* FIX: NEWS: Fixed ancient NEWS entries. Removed umlauts
* FIX: fix macname cache, SF bug 1021642
* FIX: revert Makefile change from 2.0-rc1. We have to include BDB_CFLAGS
       after CFLAGS

Changes in 2.0-rc1
==================

* NEW: new manual page for asip-status.pl
* UPD: updated documentation
* UPD: uams: link uam_dhx_passwd.so to lcrypt before lcrypto. might help with
       MD5 passwords
* UPD: Improved BerkeleyDB detection
* UPD: sys/solaris/Makefile.in: enable 'make check', from Alexander Barton
* UPD: tcp wrappers detection should work on OpenBSD as well now
* UPD: macbin: increase the maximum size of macbinary forks, as suggested by 
       Sourceforge bug ID 829221
* UPD: ASP: rework getstatus. use several ASP packets if the client allows
       it, otherwise just send as much as we can
* FIX: FreeBSD 5 build, from Alex Barton (alex at barton.de)
* FIX: OSX 10.3 build
* FIX: papd: workaround a problem with PJL before Postscript
* FIX: afpd: make sure we only disconnect on old session if the users match
* FIX: apfd: Quark6 mangled long filenames should work better now
* FIX: enhance ADv1 to ADv2 conversion. Fixed a SIGSEGV reported by Mark Baker
* FIX: better detection of invalid resource forks
* FIX: fix some linking problems on OpenBSD
* FIX: afpd: catsearch.c, filedir.c: fix bogus casts, from Olaf Hering 
       (olh at suse.de)
* FIX: afpd: don't try to create special folders and .volinfo on read-only
       volumes
* FIX: iconv/unicode enhancements. fixed a sigsegv on conversion error
* FIX: configure.in: fix a typo, reported by Joerg Rossdeutscher
* FIX: uniconv: enhanced uniconv behaviour
* FIX: fixed some Solaris compilation problems
* FIX: papd/Makefile.am: add a missing $DESTDIR, from Vlad Agranovsky
* FIX: afpd: quota.c: remove a c99 declaration, from Yann Rouillard
* FIX: configure.in: Solaris/gcc 3.0 fix, from Yann Rouillard
* FIX: afpd: fix a SIGSEGV when sharing home dirs without any options in 
       AppleVolumes.
* FIX: numerous small bugfixes

Changes in 2.0-beta2
====================

* NEW: atalkd, papd and npb tools now support nbpnames with extended
       characters
* NEW: integrated CUPS support for papd
* NEW: optionally advertise SSH tunneling capabilties
* NEW: automatic logfile removal for cnid_metad
* NEW: asip-status.pl has been added to netatalk
* UPD: updated documentation
* UPD: we now require Berkeley DB >= 4.1
* UPD: 64bit Linux fixes from Stew Benedict, Mandrakesoft
* UPD: remove --enable-sendfile
* UPD: more verbose error messages
* FIX: better handling for resource forks without read access
* FIX: Tru64 build, by Burkhard Schmidt
* FIX: MIT Kerberos detection
* FIX: varios *BSD compile problems
* FIX: compile problem with libiconv, reported by Joe Marcus Clarke
* FIX: adv1tov2: make it work with the new structure
* FIX: afpd: filenames longer than 127 bytes were not enumerated correctly,
       reported by Thies C. Arntzen
* FIX: afpd: return IP before FQDN in status reply.
* FIX: afpd: Mac chooser could crash on a codepage conversion error
* FIX: afpd: KerberosV auth with Panther clients, make long AD tickets work,
       reported by Andrew Smith
* FIX: atalkd: could send invalid NBPLKUP replies, e.g with more than 15
       printers. Reported by Almacha
* FIX: papd: fix papd.conf parsing problems with consecutive ':' and missing 
       newline. Reported by Craig White.
* FIX: megatron: make megatron work with UTF-8 volumes
* FIX: timeout: running timeout with commands which accept arguments, 
       from Yuval Yeret.
* FIX: uniconv: fix a SEGFAULT, reported by Matthew Geier
* FIX: pam detection: PAM_C/LDFLAGS were always empty, from Alexander Barton
* FIX: numerous small bugfixes.

Changes in 2.0-beta1
=====================

* NEW: OSX style adouble scheme
* NEW: japanese SHIFT_JIS codepage (iconv supplied)
* NEW: Solaris kernel module build integrated with configure
* NEW: Gentoo start scripts
* NEW: cnid_dbd doesn't use transactions by default
* FIX: afpd: the volume casefold option was broken
* FIX: afpd: update AD2 headers and keep owner on file exchange
* FIX: Solaris 9 and FreeBSD 4.9/5.2 compilation
* FIX: free space reported with groups quotas on Linux
* FIX: OS9/OS X didn't update free space
* FIX: finder crash if folder opened got deleted by another process
* FIX: randnum UAM wasn't AFP3 ready
* FIX: numerous small bugfixes.

Changes in 2.0-alpha2
=====================

* NEW: uniconv tool for converting volume encoding.
* NEW: afpd: Make sure getstatus doesn't return loopback address as server IP.

contrib/printing/timeout.c
/*
 * $Id: timeout.c,v 1.5 2005-04-28 20:49:36 bfernhomberg Exp $
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <ctype.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif /* HAVE_UNISTD_H */

char **av;

struct slist {
	char *name;
	int num;
} sigs[] = {
#ifdef SIGHUP
	{ "HUP",	SIGHUP },
#endif
#ifdef SIGINT
	{ "INT",	SIGINT },
#endif
#ifdef SIGQUIT
	{ "QUIT",	SIGQUIT },
#endif
#ifdef SIGILL
	{ "ILL",	SIGILL },
#endif
#ifdef SIGTRAP
	{ "TRAP",	SIGTRAP },
#endif
#ifdef SIGABRT
	{ "ABRT",	SIGABRT },
#endif
#ifdef SIGIOT
	{ "IOT",	SIGIOT },
#endif
#ifdef SIGEMT
	{ "EMT",	SIGEMT },
#endif
#ifdef SIGFPE
	{ "FPE",	SIGFPE },
#endif
#ifdef SIGKILL
	{ "KILL",	SIGKILL },
#endif
#ifdef SIGBUS
	{ "BUS",	SIGBUS },
#endif
#ifdef SIGSEGV
	{ "SEGV",	SIGSEGV },
#endif
#ifdef SIGSYS
	{ "SYS",	SIGSYS },
#endif
#ifdef SIGPIPE
	{ "PIPE",	SIGPIPE },
#endif
#ifdef SIGALRM
	{ "ALRM",	SIGALRM },
#endif
#ifdef SIGTERM
	{ "TERM",	SIGTERM },
#endif
#ifdef SIGURG
	{ "URG",	SIGURG },
#endif
#ifdef SIGSTOP
	{ "STOP",	SIGSTOP },
#endif
#ifdef SIGTSTP
	{ "TSTP",	SIGTSTP },
#endif
#ifdef SIGCONT
	{ "CONT",	SIGCONT },
#endif
#ifdef SIGCHLD
	{ "CHLD",	SIGCHLD },
#endif
#ifdef SIGCLD
	{ "CLD",	SIGCLD },
#endif
#ifdef SIGTTIN
	{ "TTIN",	SIGTTIN },
#endif
#ifdef SIGTTOU
	{ "TTOU",	SIGTTOU },
#endif
#ifdef SIGIO
	{ "IO",		SIGIO },
#endif
#ifdef SIGXCPU
	{ "XCPU",	SIGXCPU },
#endif
#ifdef SIGXFSZ
	{ "XFSZ",	SIGXFSZ },
#endif
#ifdef SIGVTALRM
	{ "VTALRM",	SIGVTALRM },
#endif
#ifdef SIGPROF
	{ "PROF",	SIGPROF },
#endif
#ifdef SIGWINCH
	{ "WINCH",	SIGWINCH },
#endif
#ifdef SIGINFO
	{ "INFO",	SIGINFO },
#endif
#ifdef SIGUSR1
	{ "USR1",	SIGUSR1 },
#endif
#ifdef SIGUSR2
	{ "USR2",	SIGUSR2 },
#endif
#ifdef SIGPWR
	{ "PWR",	SIGPWR },
#endif
	{ 0,		0 }
};

void
usage()
{
	int i;

	fprintf (stderr, "Usage: %s [-s signal] seconds program [args]\n\n",
		 *av);
	fprintf (stderr, "You can use a numerical signal, or one of these:\n");

#define COLS 8

	for (i = 0; sigs[i].name; i++) {
		if (i % COLS == 0)
			fprintf (stderr, "\n\t");

		fprintf (stderr, "%s", sigs[i].name);

		if ((i + 1) % COLS != 0)
			fprintf (stderr, "\t");
	}

	fprintf (stderr, "\n\n");
}

int
main (argc, argv)
	int argc;
	char **argv;
{
	int i;
	int whatsig = SIGTERM;
	extern char *optarg;
	extern int optind;
	int pid;

	av = argv;

	while ((i = getopt (argc, argv, "+s:")) != -1) {

		switch (i) {
		case 's':
			if (isdigit (*optarg)) {
				whatsig = atoi (optarg);
			} else {
				for (i = 0; sigs[i].name; i++) {
					if (!strcmp (sigs[i].name, optarg)) {
						whatsig = sigs[i].num;
						break;
					}
				}

				if (!sigs[i].name) {
					fprintf (stderr, 
						 "%s: Unknown signal %s\n",
						 *av, optarg);
					usage();
					exit (1);
				}
			}

			break;

		default:
			usage();
			exit (1);
		}
	}

	if (optind > argc - 2) {
		usage();
		exit (1);
	}

	if (!isdigit (*argv[optind])) {
		fprintf (stderr, "%s: \"seconds\" must be numeric, not %s\n",
			 *av, argv[optind]);
		usage();
		exit (1);
	}

	pid = fork();

	if (pid < 0) {
		fprintf (stderr, "%s: fork failed: ", *av);
		perror ("fork");
		exit (1);
	} else if (pid == 0) {
		int seconds = atoi (argv[optind]);

		pid = getppid();

		fclose (stdin);
		fclose (stdout);
		fclose (stderr);

		while (seconds-- > 0) {
			/*
			 * too bad there's no SIGPARENT so we have to keep
			 * polling to find out if it's still there
			 */

			if (kill (pid, 0) != 0)
				exit (0);

			sleep (1);
		}

		kill (pid, whatsig);
		exit (0);
	} else {
		execvp (argv[optind + 1], argv + optind + 1);

		fprintf (stderr, "%s: can't execute ", *av);
		perror (argv[optind + 1]);
		exit (1);
	}
}
